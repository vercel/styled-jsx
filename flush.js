module.exports = require('./dist/flush')

