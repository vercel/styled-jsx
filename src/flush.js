// Ours
import memory from './memory'

export default function flush() {
  const ret = {}

  for (const i in memory) {
    if (!{}.hasOwnProperty.call(memory, i)) {
      continue
    }

    ret[i] = memory[i]
    delete memory[i]
  }

  return ret
}
