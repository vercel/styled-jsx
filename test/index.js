// Packages
import test from 'ava'
import React from 'react'
import ReactDOM from 'react-dom/server'

// Ours
import plugin from '../src/babel'
import JSXStyle from '../src/style'
import flush, { flushToHTML } from '../src/server'
import _transform from './_transform'

const transform = (file, opts = {}) =>
  _transform(file, {
    plugins: [plugin],
    ...opts
  })

test('works with stateless', async t => {
  const { code } = await transform('./fixtures/stateless.js')
  t.snapshot(code)
})

test('ignores whitespace around expression container', async t => {
  const { code } = await transform('./fixtures/whitespace.js')
  t.snapshot(code)
})

test('works with class', async t => {
  const { code } = await transform('./fixtures/class.js')
  t.snapshot(code)
})

test('ignores when attribute is absent', async t => {
  const { code } = await transform('./fixtures/absent.js')
  t.snapshot(code)
})

test('works with global styles', async t => {
  const { code } = await transform('./fixtures/global.js')
  t.snapshot(code)
})

test('generates source maps', async t => {
  const { code } = await transform('./fixtures/source-maps.js', {
    sourceMaps: true
  })
  t.snapshot(code)
})

test('mixed global and scoped', async t => {
  const { code } = await transform('./fixtures/mixed-global-scoped.js')
  t.snapshot(code)
})

test('works with multiple jsx blocks', async t => {
  const { code } = await transform('./fixtures/multiple-jsx.js')
  t.snapshot(code)
})

test('should not add the data-jsx attribute to components instances', async t => {
  const { code } = await transform('./fixtures/component-attribute.js')
  t.snapshot(code)
})

test('works with expressions in template literals', async t => {
  const { code } = await transform('./fixtures/expressions.js')
  t.snapshot(code)
})

test('should have different jsx ids', async t => {
  const { code } = await transform('./fixtures/different-jsx-ids.js')
  t.snapshot(code)
})

test('works with non styled-jsx styles', async t => {
  const { code } = await transform('./fixtures/non-styled-jsx-style.js')
  t.snapshot(code)
})

test('works with css tagged template literals in the same file', async t => {
  const { code } = await transform('./fixtures/css-tag-same-file.js')
  t.snapshot(code)
})

test('does not transpile nested style tags', async t => {
  const { message } = await t.throws(
    transform('./fixtures/nested-style-tags.js')
  )
  t.regex(message, /detected nested style tag/i)
})

test('server rendering', t => {
  function App() {
    const color = 'green'
    return React.createElement(
      'div',
      null,
      React.createElement(JSXStyle, {
        css: 'p { color: red }',
        styleId: 1
      }),
      React.createElement(JSXStyle, {
        css: 'div { color: blue }',
        styleId: 2
      }),
      React.createElement(JSXStyle, {
        css: `div { color: ${color} }`,
        styleId: 3
      })
    )
  }
  // Expected CSS
  const expected =
    '<style id="__jsx-1">p { color: red }</style>' +
    '<style id="__jsx-2">div { color: blue }</style>' +
    '<style id="__jsx-3">div { color: green }</style>'

  // Render using react
  ReactDOM.renderToString(React.createElement(App))
  const html = ReactDOM.renderToStaticMarkup(
    React.createElement('head', null, flush())
  )

  t.is(html, `<head>${expected}</head>`)

  // Assert that memory is empty
  t.is(0, flush().length)
  t.is('', flushToHTML())

  // Render to html again
  ReactDOM.renderToString(React.createElement(App))
  t.is(expected, flushToHTML())

  // Assert that memory is empty
  t.is(0, flush().length)
  t.is('', flushToHTML())
})
