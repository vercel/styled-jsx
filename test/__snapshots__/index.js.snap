// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generates source maps 1`] = `
"import _JSXStyle from 'styled-jsx/style';
export default (() => <div className={\\"jsx-2743241663\\"}>
    <p className={\\"jsx-2743241663\\"}>test</p>
    <p className={\\"jsx-2743241663\\"}>woot</p>
    <_JSXStyle styleId={\\"2743241663\\"} css={\\"p.jsx-2743241663{color:red;}\\\\n/*# sourceMappingURL=data:application/json;charset=utf-8;base64,eyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbInNvdXJjZS1tYXBzLmpzIl0sIm5hbWVzIjpbXSwibWFwcGluZ3MiOiJBQUlnQixBQUNjLFVBQUMiLCJmaWxlIjoic291cmNlLW1hcHMuanMiLCJzb3VyY2VzQ29udGVudCI6WyJleHBvcnQgZGVmYXVsdCAoKSA9PiAoXG4gIDxkaXY+XG4gICAgPHA+dGVzdDwvcD5cbiAgICA8cD53b290PC9wPlxuICAgIDxzdHlsZSBqc3g+eydwIHsgY29sb3I6IHJlZCB9J308L3N0eWxlPlxuICA8L2Rpdj5cbilcbiJdfQ== */\\\\n/*@ sourceURL=source-maps.js */\\"} />
  </div>);"
`;

exports[`ignores when attribute is absent 1`] = `
"const a = () => <div>
    <p>hi</p>
    <style>{'woot'}</style>
  </div>;"
`;

exports[`ignores whitespace around expression container 1`] = `
"import _JSXStyle from 'styled-jsx/style';
export default (() => <div className={\\"jsx-2743241663\\"}>
    <p className={\\"jsx-2743241663\\"}>test</p>
    <p className={\\"jsx-2743241663\\"}>woot</p>
    <p className={\\"jsx-2743241663\\"}>woot</p>
    <_JSXStyle styleId={\\"2743241663\\"} css={\\"p.jsx-2743241663{color:red;}\\"} />
  </div>);"
`;

exports[`mixed global and scoped 1`] = `
"import _JSXStyle from 'styled-jsx/style';
const Test = () => <_JSXStyle styleId={\\"2743241663\\"} css={\\"p{color:red;}\\"} />;

export default (() => <div className={\\"jsx-2673076688\\"}>
    <p className={\\"jsx-2673076688\\"}>test</p>
    <_JSXStyle styleId={\\"4269072806\\"} css={\\"body{background:red;}\\"} />
    <_JSXStyle styleId={\\"2743241663\\"} css={\\"p.jsx-2673076688{color:red;}\\"} />
  </div>);"
`;

exports[`should have different jsx ids 1`] = `
"import _JSXStyle from 'styled-jsx/style';
const color = 'red';
const otherColor = 'green';

const A = () => <div className={\\"jsx-57381496\\"}>
    <p className={\\"jsx-57381496\\"}>test</p>
    <_JSXStyle styleId={\\"57381496\\"} css={\`p.jsx-57381496{color:\${color};}\`} />
  </div>;

const B = () => <div className={\\"jsx-3099245642\\"}>
    <p className={\\"jsx-3099245642\\"}>test</p>
    <_JSXStyle styleId={\\"3099245642\\"} css={\`p.jsx-3099245642{color:\${otherColor};}\`} />
  </div>;

export default (() => <div>
    <A />
    <B />
  </div>);"
`;

exports[`should not add the data-jsx attribute to components instances 1`] = `
"import _JSXStyle from \\"styled-jsx/style\\";
const Test = () => <div className={\\"jsx-2529315885\\"}>
    <span className={\\"jsx-2529315885\\"}>test</span>
    <Component />
    <_JSXStyle styleId={\\"2529315885\\"} css={\\"span.jsx-2529315885{color:red;}\\"} />
  </div>;"
`;

exports[`works with class 1`] = `
"import _JSXStyle from \\"styled-jsx/style\\";
export default class {

  render() {
    return <div className={\\"jsx-2101845350\\"}>
        <p className={\\"jsx-2101845350\\"}>test</p>
        <_JSXStyle styleId={\\"2101845350\\"} css={\\"p.jsx-2101845350{color:red;}\\"} />
      </div>;
  }

}"
`;

exports[`works with css tagged template literals in the same file 1`] = `
"import _JSXStyle from 'styled-jsx/style';


export default (({ children }) => <div className={\`jsx-\${styles.__scopedHash}\`}>
    <p className={\`jsx-\${styles.__scopedHash}\`}>{children}</p>
    <_JSXStyle styleId={styles.__scopedHash} css={styles.__scoped} />
  </div>);

const styles = new String('p.jsx-2587355013{color:red;}');

styles.__hash = '2587355013';
class Test extends React.Component {
  render() {
    return <div className={\`jsx-\${styles.__scopedHash}\`}>
        <p className={\`jsx-\${styles.__scopedHash}\`}>{this.props.children}</p>
        <_JSXStyle styleId={styles.__scopedHash} css={styles.__scoped} />
      </div>;
  }
}"
`;

exports[`works with expressions in template literals 1`] = `
"import _JSXStyle from 'styled-jsx/style';
const darken = c => c;
const color = 'red';
const otherColor = 'green';
const mediumScreen = '680px';
const animationDuration = '200ms';
const animationName = 'my-cool-animation';
const obj = { display: 'block' };

export default (({ display }) => <div className={'jsx-3922596756 ' + _JSXStyle.dynamic([['1795062918', [display ? 'block' : 'none']]])}>
    <p className={'jsx-3922596756 ' + _JSXStyle.dynamic([['1795062918', [display ? 'block' : 'none']]])}>test</p>
    <_JSXStyle styleId={\\"1003380713\\"} css={\`p.\${color}.jsx-3922596756{color:\${otherColor};display:\${obj.display};}\`} />
    <_JSXStyle styleId={\\"2743241663\\"} css={\\"p.jsx-3922596756{color:red;}\\"} />
    <_JSXStyle styleId={\\"602592955\\"} css={\`body{background:\${color};}\`} />
    <_JSXStyle styleId={\\"602592955\\"} css={\`body{background:\${color};}\`} />
    <_JSXStyle styleId={\\"128557999\\"} css={\`p.jsx-3922596756{color:\${color};}\`} />
    <_JSXStyle styleId={\\"128557999\\"} css={\`p.jsx-3922596756{color:\${color};}\`} />
    <_JSXStyle styleId={\\"2622100973\\"} css={\`p.jsx-3922596756{color:\${darken(color)};}\`} />
    <_JSXStyle styleId={\\"1167419394\\"} css={\`p.jsx-3922596756{color:\${darken(color) + 2};}\`} />
    <_JSXStyle styleId={\\"4052509549\\"} css={\`@media (min-width:\${mediumScreen}){p.jsx-3922596756{color:green;}p.jsx-3922596756{color:\${\`red\`};}}p.jsx-3922596756{color:red;}\`} />
    <_JSXStyle styleId={\\"2824547816\\"} css={\`p.jsx-3922596756{-webkit-animation-duration:\${animationDuration};animation-duration:\${animationDuration};}\`} />
    <_JSXStyle styleId={\\"417951176\\"} css={\`p.jsx-3922596756{-webkit-animation:\${animationDuration} forwards \${animationName};animation:\${animationDuration} forwards \${animationName};}div.jsx-3922596756{background:\${color};}\`} />

    <_JSXStyle styleId={\\"1795062918\\"} css={\`span.__jsx-style-dynamic-selector{display:\${display ? 'block' : 'none'};}\`} dynamic={[display ? 'block' : 'none']} />
  </div>);"
`;

exports[`works with global styles 1`] = `
"import _JSXStyle from 'styled-jsx/style';
const Test = () => <div className={\\"jsx-2209073070\\"}>
    <_JSXStyle styleId={\\"2209073070\\"} css={\\"body{color:red;}:hover{color:red;display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;-webkit-animation:foo 1s ease-out;animation:foo 1s ease-out;}div a{display:none;}[data-test]>div{color:red;}\\"} />
  </div>;

const Test2 = () => <_JSXStyle styleId={\\"2743241663\\"} css={\\"p{color:red;}\\"} />;"
`;

exports[`works with multiple jsx blocks 1`] = `
"import _JSXStyle from \\"styled-jsx/style\\";
const attrs = {
  id: 'test'
};

const Test1 = () => <div className={\\"jsx-2529315885\\"}>
    <span {...attrs} data-test=\\"test\\" className={\\"jsx-2529315885\\" + \\" \\" + (attrs.className != null && attrs.className || \\"\\")}>test</span>
    <Component />
    <_JSXStyle styleId={\\"2529315885\\"} css={\\"span.jsx-2529315885{color:red;}\\"} />
  </div>;

const Test2 = () => <span>test</span>;

const Test3 = () => <div className={\\"jsx-2529315885\\"}>
    <span className={\\"jsx-2529315885\\"}>test</span>
    <_JSXStyle styleId={\\"2529315885\\"} css={\\"span.jsx-2529315885{color:red;}\\"} />
  </div>;

export default class {
  render() {
    return <div className={\\"jsx-2101845350\\"}>
        <p className={\\"jsx-2101845350\\"}>test</p>
        <_JSXStyle styleId={\\"2101845350\\"} css={\\"p.jsx-2101845350{color:red;}\\"} />
      </div>;
  }
}"
`;

exports[`works with non styled-jsx styles 1`] = `
"import _JSXStyle from 'styled-jsx/style';
export default (() => <div className={\\"jsx-2743241663\\"}>
    <p className={\\"jsx-2743241663\\"}>woot</p>
    <style dangerouslySetInnerHTML={{ __html: \`body { margin: 0; }\` }}></style>
    <_JSXStyle styleId={\\"2743241663\\"} css={\\"p.jsx-2743241663{color:red;}\\"} />
  </div>);"
`;

exports[`works with stateless 1`] = `
"import _JSXStyle from 'styled-jsx/style';
export default (() => <div className={\\"jsx-2743241663\\"}>
    <p className={\\"jsx-2743241663\\"}>test</p>
    <p className={\\"jsx-2743241663\\"}>woot</p>
    <p className={\\"jsx-2743241663\\"}>woot</p>
    <_JSXStyle styleId={\\"2743241663\\"} css={\\"p.jsx-2743241663{color:red;}\\"} />
  </div>);"
`;
